File: apps/mobile/app/src/main/java/com/interactivehouse/mobile/data/model/Device.kt
package com.interactivehouse.mobile.data.model

import com.google.gson.annotations.SerializedName

data class Device(
    @SerializedName("device_uuid")
    val deviceUuid: String,

    val name: String? = null,
    val room: String? = null,
    val type: String? = "unknown",

    val capabilities: Map<String, CapabilitySpec>? = emptyMap(),
    val state: Map<String, Any>? = emptyMap(),

    ) {
    val stateOrEmpty: Map<String, Any>
        get() = state.orEmpty()

    val capabilitiesOrEmpty: Map<String, CapabilitySpec>
        get() = capabilities.orEmpty()

    val isPoweredOn: Boolean
        get() {
            val hasPowerCapability = capabilitiesOrEmpty.containsKey("power")
            return if (hasPowerCapability) {
                (stateOrEmpty["power"] as? Boolean) == true
            } else {
                // active by default, if device doesn't have a power switch
                true
            }
        }

}

data class CapabilitySpec(
    val type: String,
    val writable: Boolean,
    val min: Int? = null,
    val max: Int? = null
)

File: apps/web/src/types/device.ts
export type Device = {
  device_uuid: string;
  name: string;
  room: string;
  type: string;

  capabilities?: {
    power?: {
      type: string;
      writable: boolean;
    };
    brightness?: {
      type: string;
      min: number;
      max: number;
      writable: boolean;
    };
  };

  state: {
    power: boolean;
    brightness?: number;
  };
};